modules/caddytls/leafpemloader.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2015 Matthew Holt and The Caddy Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package caddytls
+
+import (
+	"crypto/x509"
+	"fmt"
+
+	"github.com/caddyserver/caddy/v2"
+)
+
+func init() {
+	caddy.RegisterModule(LeafPEMLoader{})
+}
+
+// LeafPEMLoader loads leaf certificates by
+// decoding their PEM blocks directly. This has the advantage
+// of not needing to store them on disk at all.
+type LeafPEMLoader []LeafCertPEM
+
+// Provision implements caddy.Provisioner.
+func (pl LeafPEMLoader) Provision(ctx caddy.Context) error {
+	repl, ok := ctx.Value(caddy.ReplacerCtxKey).(*caddy.Replacer)
+	if !ok {
+		repl = caddy.NewReplacer()
+	}
+	for k, pair := range pl {
+		for i, tag := range pair.Tags {
+			pair.Tags[i] = repl.ReplaceKnown(tag, "")
+		}
+		pl[k] = LeafCertPEM{
+			CertificatePEM: repl.ReplaceKnown(pair.CertificatePEM, ""),
+			Tags:           pair.Tags,
+		}
+	}
+	return nil
+}
+
+// CaddyModule returns the Caddy module information.
+func (LeafPEMLoader) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  "tls.leaf_cert_loader.pem",
+		New: func() caddy.Module { return new(LeafPEMLoader) },
+	}
+}
+
+// CertKeyPEMPair pairs certificate and key PEM blocks.
+type LeafCertPEM struct {
+	// The leaf certificate in PEM format.
+	CertificatePEM string `json:"certificate"`
+
+	// Arbitrary values to associate with this certificate.
+	// Can be useful when you want to select a particular
+	// certificate when there may be multiple valid candidates.
+	Tags []string `json:"tags,omitempty"`
+}
+
+// LoadLeafCertificates returns the certificates contained in pl.
+func (pl LeafPEMLoader) LoadLeafCertificates() ([]*x509.Certificate, error) {
+	certs := make([]*x509.Certificate, 0, len(pl))
+	for i, pair := range pl {
+		cert, err := x509.ParseCertificate([]byte(pair.CertificatePEM))
```

Comment:
The `x509.ParseCertificate` expects DER, not PEM ([docs](https://pkg.go.dev/crypto/x509#ParseCertificate)). This will always fail. See the logic here on how to convert PEM to DER for parsing:

https://github.com/caddyserver/caddy/blob/e1aa862e6a951d4dc9c0e89b6d42f7fea1c5aba7/modules/caddytls/connpolicy.go#L512-L533